src/@polymer/lit-element.ts

```diff
@@ -12,53 +12,167 @@
 import { TemplateResult } from '../../lit-html/lit-html.js';
 import { render } from '../../lit-html/lib/lit-extended.js';
 
-export { html } from '../../lit-html/lit-html.js';
+export { html } from '../../lit-html/lib/lit-extended.js';
+
+/**
+ * Renders attributes to the given element based on the `attrInfo` object where
+ * boolean values are added/removed as attributes.
+ * @param {*} element Element on which to set attributes.
+ * @param {*} attrInfo Object describing attributes.
+ */
+export function renderAttributes(element: HTMLElement, attrInfo: any) {
+  for (const a in attrInfo) {
+    const v = attrInfo[a] === true ? '' : attrInfo[a];
+    if (v || v === '' || v === 0)  {
+      if (element.getAttribute(a) !== v) {
+        element.setAttribute(a, v);
+      }
+    } else if (element.hasAttribute(a)) {
+      element.removeAttribute(a);
+    }
+  }
+}
+
+/**
+ * Returns a string of css class names formed by taking the properties
+ * in the `classInfo` object and appending the property name to the string of
+ * class names if the property value is truthy.
+ * @param {*} classInfo
+ */
+export function classString(classInfo: any) {
+  const o = [];
+  for (const name in classInfo) {
+    const v = classInfo[name];
+    if (v) {
+      o.push(name);
+    }
+  }
+  return o.join(' ');
+}
+
+/**
+ * Returns a css style string formed by taking the properties in the `styleInfo`
+ * object and appending the property name (dash-cased) colon the
+ * property value. Properties are separated by a semi-colon.
+ * @param {*} styleInfo
+ */
+export function styleString(styleInfo: any) {
+  const o = [];
+  for (const name in styleInfo) {
+    const v = styleInfo[name];
+    if (v || v === 0) {
+      o.push(`${camelToDashCase(name)}: ${v}`);
+    }
+  }
+  return o.join('; ');
+}
+
 export class LitElement extends PropertiesMixin(HTMLElement) {
 
-  _nextRendered: Promise<any>|null = null;
-  _nextRenderedResolver: Function|null = null;
+  private __renderComplete: Promise<any>|null = null;
+  private __resolveRenderComplete: Function|null = null;
+  private __isInvalid: Boolean = false;
 
-  ready() {
+  protected ready() {
     this.attachShadow({mode: 'open'});
     super.ready();
   }
 
-  _flushProperties() {
-    super._flushProperties();
-    // TODO(sorvell): propertiesChanged should have `_getData`
-    const result = this.render(this.__data);
+  /**
+   * Override which always returns true so that `_propertiesChanged`
+   * is called whenver properties are invalidated. This ensures `render`
+   * is always called in response to `invalidate`.
+   * @param {*} _props Current element properties
+   * @param {*} _changedProps Changing element properties
+   * @param {*} _prevProps Previous element properties
+   * @returns {boolean} Default implementation always returns true.
+   */
+  protected _shouldPropertiesChange(_props: any, _changedProps: any, _prevProps: any) {
+    return true;
```

Comment:
This is fine, but just noted there's a closure warning on master where the default `_shouldPropertiesChange` returns `changedProps` instead of casting to a `Boolean`. Was going to ask to fix that here but just realized that it's in a separate repo. :(

Reply:
Bug in Polymer. We can fix separately.